spec/errors.md

```diff
@@ -241,49 +241,73 @@ or for private implementation use that is not supported by the current implement
 > * {{The value is not one.}}
 > ```
 
-### Invalid Expression
+### Unsupported Statement
+
+An **_<dfn>Unsupported Statement</dfn>_** error occurs when a message includes a _reserved statement_.
+
+> For example, attempting to format this message
+> would always result in an _Unsupported Statement_ error:
+>
+> ```
+> .some {|horse|}
+> {{The message body}}
+> ```
 
-An **_<dfn>Invalid Expression</dfn>_** error occurs when a _message_ includes an _expression_
+## Message Function Errors
+
+A **_<dfn>Message Function Error</dfn>_** is any error that occurs
+when calling a message function implementation
+or which depend on validation associated with a specific function.
+
+### Bad Expression
+
+A **_<dfn>Bad Expression</dfn>_** error occurs when a _message_ includes an _expression_
```

Comment:
Can you add an example here? I'm still not sure what an example would be of a "bad expression" error that isn't a "bad operand" error.

Reply:
The examples for this aren't quite so clearcut as for Bad Operand and Bad Option, but I can think of at least the following:

> Consider a selection-only function `:platform` that takes no operand or options, and matches on identifiers like `macos` or `linux` for the current OS. A Bad Expression error could be emitted if its value is used as a markup option value, as in:
> ```
> .local $foo = {:platform}
> {{{#bar opt=$foo}}}
> ```

OTOH, a bare Message Function Error could be valid for that too, and for a number of other cases that for whatever reason don't exactly fit within the categories we define here. Perhaps it'd be better to drop Bad Expression and make it explicitly clear that this list is not exhaustive?

Reply:
> Perhaps it'd be better to drop Bad Expression and make it explicitly clear that this list is not exhaustive?

We need to make an exhaustive list, from the point of view that every error is categorized. I'm okay with there being an error that is the moral equivalent of a plain `Exception` or that (as we currently have) there are some overarching errors (wasn't that what Message Function Error was going to be??). Implementers shouldn't have to guess what type of error to emit ("should I emit a Formatting Error or a Bad Expression or a Resolution Error?") even if on some level it doesn't matter.

I think I would leave out Bad Expression, especially since the things that trigger it are not "badness" in the expression but rather "badness" in the function's operation.